src/libultra/os/initialize.c

```diff
@@ -59,13 +60,49 @@ void OSINITIALIZE_FUNC(void) {
     __osSetWatchLo(0x04900000);
 #endif
 
-    while (__osSiRawReadIo((void*)(PIF_RAM_END - 3), &pifdata)) {
-        ;
+#ifdef BBPLAYER
+    {
+        u32 x, y;
+
+        // Check for iQue Player hardware by enabling and disabling FLASH and IDE interrupts and checking if the
+        // register gives the correct response.
+        IO_WRITE(MI_EX_INTR_MASK_REG, MI_EX_INTR_MASK_SET_FLASH | MI_EX_INTR_MASK_SET_IDE);
+        x = IO_READ(MI_EX_INTR_MASK_REG);
+        IO_WRITE(MI_EX_INTR_MASK_REG, MI_EX_INTR_MASK_CLR_FLASH | MI_EX_INTR_MASK_CLR_IDE);
+        y = IO_READ(MI_EX_INTR_MASK_REG);
+
+        __osBbIsBb =
+            ((x & (MI_EX_INTR_MASK_FLASH | MI_EX_INTR_MASK_IDE)) == (MI_EX_INTR_MASK_FLASH | MI_EX_INTR_MASK_IDE)) &&
+            ((y & (MI_EX_INTR_MASK_FLASH | MI_EX_INTR_MASK_IDE)) == 0);
     }
-    while (__osSiRawWriteIo((void*)(PIF_RAM_END - 3), pifdata | 8)) {
-        ;
+
+    //! @bug Most games do not have permission to use GPIO so often cannot correctly tell if they are HW V1 or V2.
+    if (__osBbIsBb && PI_GPIO_IS_HW_V2(IO_READ(PI_GPIO_REG))) {
```

Comment:
"use GPIO so often" seems like its missing a word or a comma

Reply:
sorry even with a comma this one still doesnt really make sense to me. what is the thing that cannot tell?

Reply:
Games can't tell because they don't have permission

Reply:
maybe just adding the word "they" fixes my problem, "so they often cannot tell"